STYLE.md

```diff
@@ -0,0 +1,432 @@
+```
+
+* For functions, the documentation should give enough information to
+apply the function without looking at its definition.
+
+Record example:
+
+```haskell
+-- | Bla bla bla.
+data Person = Person
+  { age  :: !Int     -- ^ Age
+  , name :: !String  -- ^ First name
+  }
+```
+
+For fields that require longer comments, format them this way:
+
+```haskell
+data Record = Record
+  { -- | This is a very very very long comment that is split over
+    -- multiple lines.
+    field1 :: !Text
+
+    -- | This is a second very very very long comment that is split
+    -- over multiple lines.
+  , field2 :: !Int
+  }
+```
+
+#### End-of-Line Comments
+
+Separate end-of-line comments from the code with 2 spaces. Align
+comments for data type definitions. Some examples:
+
+```haskell
+data Parser =
+  Parser
+    !Int         -- ^ Current position
+    !ByteString  -- ^ Remaining input
+
+foo :: Int -> Int
+foo n = salt * 32 + 9
+  where
+    salt = 453645243  -- Magic hash salt.
+```
+
+#### Links
+
+Use in-line links economically.  You are encouraged to add links for
+API names. It is not necessary to add links for all API names in a
+Haddock comment. We therefore recommend adding a link to an API name
+if:
+
+* The user might actually want to click on it for more information (in
+  your opinion), and
+
+* Only for the first occurrence of each API name in the comment (don't
+  bother repeating a link)
+
+Top-down guideline
+------------------
+
+### LANGUAGE extensions section
+
+Write each `LANGUAGE` pragma on its own line, sort them alphabetically and
+align by max width among them.
+
+```haskell
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE Rank2Types       #-}
+{-# LANGUAGE TemplateHaskell  #-}
+```
+
+### Module name
+
+Use singular when naming modules (e.g. use `Data.Map` and
+`Data.ByteString.Internal` instead of `Data.Maps` and
+`Data.ByteString.Internals`). Sometimes it's acceptable to use plural
+(e. g. `Types`, `Instances`).
+
+### Export Lists
+
+Format export lists as follows:
+
+```haskell
+module Data.Set
+  ( Set
+  , empty
+  , singleton
+  , member
+  ) where
+```
+
+Some clarifications:
+
+1. Use 2 spaces indentation for export list.
+2. You can split export list into sections or just write all as single section.
+3. It is strongly adviced to sort each section alpabetically. However,
+   classes, data types and type aliases should be written before
+   functions.
+
+### Imports
+
+Imports should be grouped in the following order:
+
+0. Implicit import of custom prelude (for example [`cardano-prelude`](https://github.com/input-output-hk/cardano-prelude)) if used.
+1. External modules.
+2. External IOHK modules.
+3. Local library modules.
+4. Local test modules
+
+* Put a blank line between each group of imports.
+
+* The imports in each group should be sorted alphabetically, by module name.
+
+* Always use explicit import lists or `qualified` imports.
+Try to use `qualified` imports only if import list is
```

Comment:
What is the rationale for discouraging `qualified` imports?

Can you explain the robustness-against-change argument (if I read it correctly that it applies)?

Reply:
`qualified` imports take away from code readability IMO. It's usually better to do an explicit import.

Reply:
So it's not about robustness-against-change, I see.

But there is another side -- explicit imports add a lot of bookkeeping cost, which qualified imports help avoid.

Also, which scenarios do you envision where qualified imports reduce readability?

I often find it the other way -- having many names used in a module can make the trip to the import section all too frequent.  Qualified names hint at which part of the realm the name comes from.  I sometimes reserve unqualified imports to names that have unambiguous interpretation and don't require this trip.

Reply:
I'd have to think about this. I typically only use qualified imports where I think they would increase readability (which includes resolving symbol conflicts), but that's quite vague.

Reply:
Reading code with a lot of qualified imports makes less readable IMO. It's difficult to define a hard line and say exactly under what conditions you should qualify an import vs explicit importation. 

I'd say use qualified imports to mostly avoid name conflicts. @erikd

Reply:
There is another angle -- patch conflicts.

Since code changes so often cause changes to the set of used names, explicit import lists increases the probability of conflicts -- relative to qualified imports.

This is because changes in the set of imports from any particular module happen more often than the change in the set of used modules -- quite naturally.

Remember how many times you were resolving those conflicts around import lists..

Reply:
The are certain libraries that make far more sense to import qualified. For instance, `Set` and `Map` which i usually use as:
```
import           Data.Map.Strict (Map)
import qualified Data.Map.Strict as Map
import           Data.Set (Set)
import qualified Data.Set as Set
```
Once they are often used like that, I even think its best to *always* use them like that. Modules that make sense to always use qualified include the strict and lazy versions of `ByteString`, `Text` and `List`.

I also agree with @deepfire 's suggestion that changes to import lists can cause changes with little to no semantic value.

Reply:
Another question -- how do we treat local (intra-package) module imports.

Do we apply the same policy of no implicit imports for them as well?

Relevant differences that can affect balance in that case:

1. Intra-package module changes happen more often, than cross-package, making explicit import list churn even worse.
1. Since the names are local, they arguably are less prone to ambiguous interpretation.
1. People seem to rarely use qualified import lists for local modules.

Reply:
> Since code changes so often cause changes to the set of used names, explicit import lists increase the probability of conflicts. This is especially bad, because the import section is rather dense, per-line change frequency -wise.
Fair enough. Should we do the following:
* Explicit or qualified imports from external packages.
@deepfire

Reply:
@Jimbo4350, frankly, I am in favor of that, but I'd want consensus between all of us.

cc @erikd, @intricate.

Reply:
I agree with @deepfire that qualified imports can reduce the PITA of explicit imports. I feel as though I've spent a lot of cumulative hours at IOHK refactoring explicit imports, massaging them around to meet line lengths (before we had auto-formatters, or when they don't work), and generally just thinking about them when I would've rather been doing actual work.

I haven't worked on a large project that used primarily qualified imports, so I don't know if there are other problems, but it seems to me that we might benefit from using more qualified and fewer explicit imports.

Reply:
As we aren't using an auto-formatter I also agree here. I was spoiled by them and therefore favoured explicit imports. I'll mark this as resolved :).